File: kotlin/word-count/src/main/kotlin/WordCount.kt
object WordCount {

    private val regex = Regex("""[A-Za-z0-9]+(?:'[A-Za-z0-9]+)*""")

    fun phrase(phrase: String): Map<String, Int> {
        val words = mutableMapOf<String, Int>()
        val input = phrase.lowercase()

        for (match in regex.findAll(input)) {
            val word = match.value
            words[word] = (words[word] ?: 0) + 1
        }

        return words
    }
}

File: csharp/word-count/WordCount.cs
using System.Text.RegularExpressions;

public static class WordCount
{
    private static Regex regex = new Regex("""[A-Za-z0-9]+(?:'[A-Za-z0-9]+)*""");

    public static IDictionary<string, int> CountWords(string phrase)
    {
        var words = new Dictionary<string, int>();
        var input = phrase.ToLower();

        foreach (Match match in regex.Matches(input))
        {
            var word = match.Value;

            if (!words.TryAdd(word, 1))
            {
                words[word]++;
            }
        }

        return words;
    }
}